vllm/worker/model_runner.py

```diff
@@ -527,30 +528,31 @@ def capture_model(self, kv_caches: List[KVCache]) -> None:
 
         # NOTE: Capturing the largest batch size first may help reduce the
         # memory usage of CUDA graph.
-        for batch_size in reversed(batch_size_capture_list):
-            # Create dummy input_metadata.
-            input_metadata = InputMetadata(
-                is_prompt=False,
-                slot_mapping=slot_mapping[:batch_size],
-                prompt_lens=None,
-                max_seq_len=None,
-                start_loc=None,
-                max_context_len=self.max_context_len_to_capture,
-                context_lens=context_lens[:batch_size],
-                block_tables=block_tables[:batch_size],
-                use_cuda_graph=True,
-            )
-
-            graph_runner = CUDAGraphRunner(self.model)
-            graph_runner.capture(
-                input_tokens[:batch_size],
-                input_positions[:batch_size],
-                kv_caches,
-                input_metadata,
-                memory_pool=self.graph_memory_pool,
-            )
-            self.graph_memory_pool = graph_runner.graph.pool()
-            self.graph_runners[batch_size] = graph_runner
+        with custom_all_reduce.capture():
```

Comment:
out of curiosity, is it possible to put this in `CUDAGraphRunner.capture` instead? or does this need to be a single context for all graphs?

Reply:
Yes we can. `custom_all_reduce.capture()` will work with repeated invocations.

Reply:
Actually, that will increase the cuda graph capture time from 7s to 39s. This is because batch registering all cuda graph addresses at once will be more efficient. Let me know if this change is still worth it.

Reply:
Actually I tried it with multiple captures and it wasn't working correctly (output was gibberish) :(